### PR TITLE
disttask: add option to run subtask concurrently

### LINKED ISSUE #43369: disttask: scheduler can run subtask concurrently

### PR DESCRIPTION
**Pull request by GMHDBJD**
### What problem does this PR solve?

Issue Number: close #43369

Problem Summary:

### What is changed and how it works?
- move onSubtaskFinished to minimal task pool(the last one minimal task call onSubtaskFinished)
- add option to run subtask concurrently
### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: disttask/framework/scheduler/scheduler.go [hunk 1]
```diff
@@ -116,67 +118,101 @@ func (s *InternalSchedulerImpl) Run(ctx context.Context, task *proto.Task) error
...
 		}
-		if err := s.getError(); err != nil {
-			if errors.Cause(err) == context.Canceled {
-				s.updateSubtaskStateAndError(subtask.ID, proto.TaskStateCanceled, "")
-			} else {
-				s.updateSubtaskStateAndError(subtask.ID, proto.TaskStateFailed, s.getError().Error())
-			}
-			break
+	}
+	if err := s.getError(); err != nil {
```
Comment (tangenta) [inline disttask/framework/scheduler/scheduler.go:207]:
```suggestion
	} else {
```

File: disttask/framework/scheduler/scheduler.go [hunk 2]
```diff
@@ -116,67 +118,101 @@ func (s *InternalSchedulerImpl) Run(ctx context.Context, task *proto.Task) error
...
-		for _, minimalTask := range minimalTasks {
-			minimalTaskWg.Add(1)
-			j := minimalTask
-			minimalTaskCh <- func() {
-				s.runMinimalTask(runCtx, j, task.Type, task.Step)
-				minimalTaskWg.Done()
+		s.subtaskWg.Done()
+		return
+	}
+	logutil.Logger(s.logCtx).Info("split subTask", zap.Any("cnt", len(minimalTasks)), zap.Any("subtask_id", subtask.ID))
```
Comment (tangenta) [inline disttask/framework/scheduler/scheduler.go:171]:
```suggestion
	logutil.Logger(s.logCtx).Info("split subTask", zap.Int("cnt", len(minimalTasks)), zap.Int64("subtask_id", subtask.ID))
```

File: disttask/framework/scheduler/register.go [hunk 3]
```diff
@@ -18,14 +18,23 @@ import (
...
 // Constructor is the constructor of Scheduler.
 type Constructor func(taskMeta []byte, step int64) (Scheduler, error)
 
 // RegisterOption is the register option of Scheduler.
 type RegisterOption func(opts *schedulerRegisterOptions)
 
+// WithConcurrentSubtask is the option of Scheduler to run subtasks concurrently.
+func WithConcurrentSubtask() RegisterOption {
+	return func(opts *schedulerRegisterOptions) {
+		opts.ConcurrentSubtask = true
```
[the remark this answers is not in the document]
Comment (GMHDBJD) [reply disttask/framework/scheduler/register.go:34]: wenjun said it should be false in add index. cc @wjhuang2016

**Review by lcwangchao — APPROVED** [review]

File: disttask/framework/scheduler/scheduler.go (hunk 1 above)
Comment (GMHDBJD) [reply disttask/framework/scheduler/scheduler.go:207]: if OnSubtaskFinished return error(L204), we also need to update subtask status to failed

File: disttask/framework/scheduler/scheduler.go [hunk 4]
```diff
@@ -116,67 +118,101 @@ func (s *InternalSchedulerImpl) Run(ctx context.Context, task *proto.Task) error
...
+		minimalTaskCh <- func() {
+			s.runMinimalTask(ctx, j, subtask.Type, step)
+
+			mu.Lock()
+			defer mu.Unlock()
+			cnt++
+			// last minimal task should mark subtask as finished
+			if cnt == len(minimalTasks) {
+				s.onSubtaskFinished(ctx, scheduler, subtask)
+				s.subtaskWg.Done()
```
[the remark this answers is not in the document]
Comment (GMHDBJD) [reply disttask/framework/scheduler/scheduler.go:195]: if move `onSubtaskFinished` out of `runSubtask`, which means we need to manage the goroutine of `onSubtaskFinished` in scheduler. Instead, it is now managed by the scheduler pool

**Review by tangenta — APPROVED** [review]

**Pull request merged**